mm/mmap.c

```diff
@@ -276,19 +276,21 @@ void free_vma_pages(struct vma *vma, unsigned int start, __size_t length)
 		if(pgdir[pde] & PAGE_PRESENT) {
 			pgtbl = (unsigned int *)P2V((pgdir[pde] & PAGE_MASK));
 			if(pgtbl[pte] & PAGE_PRESENT) {
-				/* make sure to not free reserved pages */
-				page = pgtbl[pte] >> PAGE_SHIFT;
-				pg = &page_table[page];
-				if(pg->flags & PAGE_RESERVED) {
-					continue;
-				}
+				if (!(pgtbl[pte] & PAGE_NOALLOC)) {
+					/* make sure to not free reserved pages */
+					page = pgtbl[pte] >> PAGE_SHIFT;
+					pg = &page_table[page];
+					if(pg->flags & PAGE_RESERVED) {
+						continue;
+					}
 
-				if(vma->prot & PROT_WRITE && vma->flags & MAP_SHARED) {
-					offset = start - vma->start + vma->offset + n * PAGE_SIZE;
-					write_page(pg, vma->inode, offset, PAGE_SIZE);
-				}
+					if(vma->prot & PROT_WRITE && vma->flags & MAP_SHARED) {
+						offset = start - vma->start + vma->offset + n * PAGE_SIZE;
+						write_page(pg, vma->inode, offset, PAGE_SIZE);
+					}
 
-				kfree(P2V(pgtbl[pte]) & PAGE_MASK);
+					kfree(P2V(pgtbl[pte]) & PAGE_MASK);
+				}
```

Comment:
Wouldn't be easier to follow the same previous block scheme with `PAGE_RESERVED` and avoid all this indentation?
``` C
				if(pgtbl[pte] & PAGE_NOALLOC) {
					continue;
				}
```

Reply:
No, because the current->rss-- needs to be decremented, and the page table entry needs to be zeroed with `pgtbl[pte] = 0;` below.

Reply:
Oh!, you're right, I should have seen the whole file.